### PR TITLE
Delimited types

### PR DESCRIPTION
**Pull request by pavel-kirienko**
This PR implements the delimited serialization logic per https://github.com/UAVCAN/specification/issues/90 as described in https://forum.uavcan.org/t/data-type-extensibility-and-composition/827/15. Please read the linked thread.

The changes described on the forum require review. The code changes are rather mundane and are probably not worth looking into, but you can look at the auto-generated docs at https://pydsdl--45.org.readthedocs.build/en/45/

Once this is merged, I will introduce the necessary changes to the DSDL chapter of the Specification.

The corresponding changes to the public regulated types are available at https://github.com/UAVCAN/public_regulated_data_types/tree/beta. There are no implications for currently existing systems leveraging v1 (compatibility not affected).

I removed the demo.py because maintaining it in a sensible state is just too much work. Instead, those looking for examples can just go look at Nunavut.

### CODE REVIEW

File: pydsdl/_bit_length_set.py [hunk 1]
```diff
@@ -283,22 +244,69 @@ def __radd__(self, other: typing.Any) -> 'BitLengthSet':
...
         BitLengthSet({5, 6, 7, 8, 9})
         """
         if isinstance(other, _OPERAND_TYPES):
-            self.increment(other)
+            self._value = (self + other)._value
+            return self
+        else:
+            return NotImplemented
+
+    def __or__(self, other: typing.Any) -> 'BitLengthSet':
```
Comment (thirtytwobits) [inline pydsdl/_bit_length_set.py:266]: image

File: pydsdl/_data_schema_builder.py [hunk 2]
```diff
@@ -53,16 +43,20 @@ def attributes(self) -> typing.List[_serializable.Attribute]:  # noinspection Py
     def empty(self) -> bool:
         return not self._fields and not self._constants
 
+    @property
+    def extent(self) -> typing.Optional[int]:
```
Comment (thirtytwobits) [inline pydsdl/_data_schema_builder.py:47]: this needs an approved update to the specification before adding.

File: pydsdl/_data_schema_builder.py [hunk 3]
```diff
@@ -53,16 +43,20 @@ def attributes(self) -> typing.List[_serializable.Attribute]:  # noinspection Py
...
+        return self._extent
+
+    @property
+    def final(self) -> bool:
```
Comment (thirtytwobits) [inline pydsdl/_data_schema_builder.py:51]: This needs an approved update to the specification before adding.

File: pydsdl/_bit_length_set.py [hunk 4]
```diff
@@ -66,63 +67,56 @@ def is_aligned_at(self, bit_length: int) -> bool:
...
-        >>> a
-        BitLengthSet({5, 6, 7, 8, 9})
-        >>> a = BitLengthSet()
-        >>> a.increment({1, 2, 3})
-        >>> a
-        BitLengthSet({1, 2, 3})
+        from ._serializable import SerializableType
+        return self.is_aligned_at(SerializableType.BITS_PER_BYTE)
+
+    def pad_to_alignment(self, bit_length: int) -> 'BitLengthSet':
```
Comment (thirtytwobits) [inline pydsdl/_bit_length_set.py:80]: All use of this method must trace to approved changes in the specification.

File: pydsdl/_data_type_builder.py [hunk 5]
```diff
@@ -60,36 +60,34 @@ def __init__(self,
...
-                    source_file_path=self._definition.file_path)  # type: _serializable.CompositeType
+            ty = _serializable.UnionType if struct.union else _serializable.StructureType
+            fin = ty(name=self._definition.full_name,
+                     version=self._definition.version,
+                     attributes=struct.attributes,
+                     deprecated=self._is_deprecated,
+                     fixed_port_id=self._definition.fixed_port_id,
+                     source_file_path=self._definition.file_path)  # type: _serializable.CompositeType
+            if not struct.final:
+                out = _serializable.DelimitedType(fin, extent=struct.extent)  # type: _serializable.CompositeType
```
Comment (thirtytwobits) [inline pydsdl/_data_type_builder.py:71]: This is a huge change to the core protocol. We need to first modify the specification.

File: pydsdl/_bit_length_set.py [hunk 6]
```diff
@@ -66,63 +67,56 @@ def is_aligned_at(self, bit_length: int) -> bool:
...
+        >>> BitLengthSet({0, 1, 5, 7}).pad_to_alignment(2)
+        BitLengthSet({0, 2, 6, 8})
+        >>> BitLengthSet({0, 1, 2, 3, 4, 5, 6, 7, 8}).pad_to_alignment(3)
+        BitLengthSet({0, 3, 6, 9})
+        >>> BitLengthSet({0, 1, 2, 3, 4, 5, 6, 7, 8}).pad_to_alignment(8)
+        BitLengthSet({0, 8})
+        >>> BitLengthSet({0, 9}).pad_to_alignment(8)
+        BitLengthSet({0, 16})
+        >>> from random import randint
+        >>> alignment = randint(1, 64)
```
[the remark this answers is not in the document]
Comment (pavel-kirienko) [reply pydsdl/_bit_length_set.py:99]: Indeed, I should use https://pypi.org/project/pytest-randomly/

File: pydsdl/_bit_length_set.py (hunk 1 above)
Comment (pavel-kirienko) [reply pydsdl/_bit_length_set.py:266]: I think they map well to the subject domain. Do you think I should define regular methods instead?

Comment (thirtytwobits) [reply pydsdl/_bit_length_set.py:266]: I'm totally fine with how you are doing this. Just having some meme fun with ya.

File: pydsdl/_bit_length_set.py [hunk 7]
```diff
@@ -66,63 +67,56 @@ def is_aligned_at(self, bit_length: int) -> bool:
...
+        >>> BitLengthSet(randint(1, 1000) for _ in range(100)).pad_to_alignment(alignment).is_aligned_at(alignment)
+        True
         """
-        self._value = BitLengthSet.elementwise_sum_cartesian_product([self or BitLengthSet(0),
-                                                                      BitLengthSet(bit_length_set_or_scalar)])._value
+        r = int(bit_length)
+        if r < 1:
+            raise ValueError('Invalid alignment: %r bits' % r)
+        elif r == 1:
+            return BitLengthSet(self)  # This is a no-op, supported for completeness.
```
[the remark this answers is not in the document]
Comment (pavel-kirienko) [reply pydsdl/_bit_length_set.py]: Indeed, that was unnecessary, removed.

File: pydsdl/_bit_length_set.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (pavel-kirienko) [reply pydsdl/_bit_length_set.py:99]: Done.

**Comment by pavel-kirienko** [comment]
@thirtytwobits should be finished now, I think.

File: pydsdl/_data_type_builder.py [hunk 8]
```diff
@@ -211,31 +230,60 @@ def _on_assert_directive(self, line_number: int, value: typing.Optional[_express
...
+
+    @staticmethod
+    def _render_explicit_extent_recommendation(model: _serializable.DelimitedType) -> str:
+        assert model.extent % 8 == 0, 'Internal error: Extent is expected to be an integer multiple of 8 bits'
+        e_bytes = model.extent // 8
+        return (
+            'Extent is not specified explicitly, defaulting to {e_bytes} bytes. '
+            'To accept the default extent and silence this diagnostic, '
+            'add the following line near the end of the definition: '
+            '`@extent {e_bytes} * 8`'
```
Comment (thirtytwobits) [inline pydsdl/_data_type_builder.py:288]: 👍

**Review by thirtytwobits — APPROVED** [review]
Super-fast review but generally looks okay.

**Pull request merged**